### ISSUE TITLE
Converted image aspec ratio

### ISSUE THREAD

**Issue by tsauerwein** [c2corg/v6_images#5]
Reported by @ginold:

> The resized uploaded images (for example the ones ending on "SI" - small image) do not hold their original aspect ratio. They are all squares, therefore they appear distorted on thumbnails.

https://github.com/c2corg/v6_api/issues/356

**Comment by asaunier** [comment]
Yes, in v5, thumbnails are cropped and squared.
2 options:
* we keep the thumbnails as squares and display them as is in the list of images
* we regenerate the thumbnails to keep the full image format.

**Comment by tsauerwein** [comment]
@asaunier I think the problem @ginold sees is that the images are not cropped. Using squares for thumbnails is ok, but we must make sure that the images are cropped correctly and not simply resized.

**Comment by ginold** [comment]
If we want to continue to crop images to a square, it will cut some part of the image. Is there a problem with resizing the image proportionally ?

**Comment by asaunier** [comment]
> If we want to continue to crop images to a square, it will cut some part of the image.

That's what is already done in v5 http://www.camptocamp.org/images

> Is there a problem with resizing the image proportionally ?

You mean keeping the orginal image proportions?
No problem to change this behaviour. But that means that all thumbnails from v5 must be regenerated in order to have consistent thumbnail formats.

**Comment by ginold** [comment]
So maybe at least let's keep the old thumbnails as they are, but the new ones should be cropped proportionally ?

**Comment by asaunier** [comment]
What does "be cropped proportionally" mean for you?
For me "crop" means that some parts of the image are stripped in order to get a final square shape. The content should not be distorted.

**Comment by ginold** [comment]
well, i should say "resize", not crop. That's what I said - if we crop, we will loose some part of the image, so can't we resize it to a size of thumbnail, keeping the propotions, aspect ratio?

**Comment by asaunier** [comment]
If it's more convenient, having thumbnails sticking to the original image format (not square as in v5) would be possible. Especially if anyway the resized images are regenerated during the v5>v6 migration. See https://github.com/c2corg/v6_images/issues/6

**Comment by tsauerwein** [comment]
If we want to keep squared thumbnails, this function would be useful:
http://pillow.readthedocs.io/en/3.1.x/reference/ImageOps.html#PIL.ImageOps.fit

**Comment by asaunier** [comment]
Discussed a few minutes ago with @ginold 
* thumbnails (SI images) don't have to square. They currently are because of the old v5 thumbnails that do are square.
* when regenerating the thumbnails, it would be an option to keep the original w/h ratio.

**Comment by arnaud-morvan** [comment]
I've finished regeneration and migration scripts in images backend. See 
I will focus on this right now, what is expected here : 
* constant height
* constant width
* cropped image with constant height and constant width

**Comment by arnaud-morvan** [comment]
As @ginold made a PR https://github.com/c2corg/v6_ui/pull/699 which need resized images to have the same w/h ratio (for animation) as BI images, I propose to conform to UI side.

We could show the result in the demo and I expect the association will simply validate the result.

I want to set this before starting the images migration.

Seems UI will need 150x150> and 900x900> (x6) (with width/height ratio from original image), I do not see where MI images are used.

@ginold, can you confirm my assumptions, and/or add comments ?

**Comment by asaunier** [comment]
> I do not see where MI images are used.

In cards

>  900x900>

You mean we would enlarge the BI images from 800px max to 900 max?
That seems a bit short. I would rather expect something like 1500px.

One problem with those BI images: they are both used in the slideshows (so the bigger the better) but also in image detail view. Not sure we have enough space to show a 1500px image in such pages? Or we could make sure the data block (with EXIF etc.) is always shown below the image, not beside it.

By the way I have a question from the association:
is it possible to resize landscape images to some max value (say 1500 or 1600px) and portrait images to some other max value (say 1200px)?

Just to make sure: original images are sometimes smaller than the max dimensions of the resized images. People sometimes upload 600px wide images (or even smaller!!) => the resized images should of course not be bigger than the originals :P

**Comment by arnaud-morvan** [comment]
BI images could be 1500x900>, according to UI animation needs, SI images should be 250x150>.
As sais I don't see where MI images are used <=> detail view ???

**Comment by arnaud-morvan** [comment]
MI images could be 900x900>

**Comment by arnaud-morvan** [comment]
Ok missed for MI in the cards.

**Comment by ginold** [comment]
> Seems UI will need 150x150> and 900x900> (x6) (with width/height ratio from original image)

x6 was just my example. It should be the exact number between SI size and BI, the number how many times smaller the SI are from the BI, so it depends on what we decide.

> I do not see where MI images are used.

to be honest - me neither :P maybe if they are around 800-900px we could use them in the detailed view.

**Comment by asaunier** [comment]
> to be honest - me neither :P maybe if they are around 800-900px we could use them in the detailed view.

You mean:
* BI (1500px) => slideshow
* MI (800-900px) => detail view
* SI => carousel AND cards? Then 150px? (not 250px as currently used in cards, perhaps a bit too big for cards anyway?) SI would then no longer be square?
By the way, don't carousel images need to have all the same height?

**Comment by arnaud-morvan** [comment]
Note that we could also have a fixed height, with adaptative width (easy on the image backend side).

**Comment by arnaud-morvan** [comment]
We can also set a fixed height and a max width using cropping, but here we may cut the image.

**Comment by asaunier** [comment]
> Note that we could also have a fixed height, with adaptative width (easy on the image backend side).

For thumbnails (SI), that would probably make sense: height=150px, width=150*original_ratio.
@ginold what do you think?

**Comment by ginold** [comment]
I think it makes sense.

> By the way, don't carousel images need to have all the same height?

no, the plugin takes care of fitting them.

**Comment by asaunier** [comment]
What about the sizes? See https://github.com/c2corg/v6_images/issues/5#issuecomment-252206572
Do we need an other image size for cards ("CI")? Or can we use the same size for cards and carousel?

**Comment by ginold** [comment]
well, I'd say that it would be best to have SI images not smaller than 200 or 250px so that we can use them in both cards and carousel without quality loss. If we want to have SI with 150px we should ask ourselves if it's ok to use them and stretch them slightly to fit to the cards (it should not be a very perceived degradation) :P

**Comment by asaunier** [comment]
* Well, I guess it's better to avoid having too many resized versions of an image. If we could use the same image for carousels and cards that would be interesting.
* Please note that in v5 there's already a kind of cards page with thumbnails (120x120): http://www.camptocamp.org/images
* I agree it's cool that the thumbnails in cards are not too small. 200px would be cool but 150px looks already not too bad to me. Here is how the v5 page looks with 150px-thumbnails:
![sans titre](https://cloud.githubusercontent.com/assets/1192331/19263312/06250e12-8f9c-11e6-8dfb-e9c52447f14f.png)
* I am a bit reluctant to artificially enlarge images by stetching them => the apparent quality of 150px-thumbnails stretched to 200px will be poor.
* What about mobile screens? How do the carousel-images and card-images behave on small screens?
* What the file weight difference between 150-px and 200px thumbnails?

**Comment by ginold** [comment]
> cool but 150px looks already not too bad to me

I would agree but there is probably a problem with the image quality - the photos look blurred and to be of poor quality, but 150x150 images on google-images look decent.  Probably something to change in the settings.

**Comment by asaunier** [comment]
> I would agree but there is probably a problem with the image quality - the photos look blurred and to be of poor quality

I agree, that's what I have also written:
> I am a bit reluctant to artificially enlarge images by stetching them

I was just wondering if using 150px-thumbnails in cards wasn't enough.

**Comment by ginold** [comment]
I would say that 150x may be just a little bit to small for using it in the cards whereas 250x just too big. 200x would fit pretty well in the cards and would be just a little too big for the carousel - it's more everything-proof.  

And surely the size is not a big deal, but I would have a hard time telling you the exact difference. Maybe @arnaud-morvan could export a few images in both sizes and compare them?

**Comment by asaunier** [comment]
@mfournier told me that it's not that the storage service is not that expensive.
So - if needed - we could consider creating a new "CI" format ("card image"), distinct from the carousel thumbnails (SI).

**Comment by asaunier** [comment]
@arnaud-morvan I think it is too early to decide what resized images we need and what their sizes are.

Because for instance it also depends on the layout of the homepage feed, that will also show images (maybe also small square ones). See https://github.com/c2corg/v6_ui/issues/625#issuecomment-253439077

**Comment by arnaud-morvan** [comment]
Ok, let me know when decided, or any question about possibilities.

The migration time mainly depends on network and S3 servers latency.

Now with #16, migration and resize scripts are multithreaded, I hope it should take less than one day for overall migration or resizing.

**Comment by asaunier** [comment]
For the beta version, I would suggest that:
* only a subset of images are migrated => to test the migration script and actually have images in the migrated documents (even though there's maybe an issue with images associated to outings? See https://github.com/c2corg/v6_ui/issues/727). We could also figure out if the quality of resized images is good.
* for image sizes, we could use temporary values:
 * BI: 1500px max on the largest side
 * MI: same as in v5 (I don't remember)
 * SI: 150px square images, preserved aspect ratio

Sounds reasonable?

**Comment by arnaud-morvan** [comment]
Note that if we migrate only a subset of images, will be difficult to find them on the demo.

**Comment by asaunier** [comment]
Most images are associated to outings. So if we migrate recent images, we should see them in recent outings. Unless c2corg/v6_ui#727

The homepage feed should also advertise recent images, see https://github.com/c2corg/v6_ui/pull/731

**Comment by tsauerwein** [comment]
> The homepage feed should also advertise recent images, see c2corg/v6_ui#731

Currently there are no images referenced in the entries of the feed.

**Comment by asaunier** [comment]
Discussion about image sizes is moved to https://github.com/c2corg/v6_images/issues/6#issuecomment-256309185

**Comment by asaunier** [comment]
Aspect ratio is not preserved in the squared thunbnails (SI) when uploading an image: SI images should be cropped instead of distorted. For instance:
![1478817982_484301537si](https://cloud.githubusercontent.com/assets/1192331/20197684/0ceb25fc-a7a1-11e6-89fb-245142045bbb.jpg) => NOK

In addition the quality of the thumbnail looks pretty poor. See https://github.com/c2corg/v6_images/issues/10

**Comment by arnaud-morvan** [comment]
This image do not seems distorded but cropped.
Same question as #10 : Where is the original image, can you give the link to the image in the demo.

**Comment by asaunier** [comment]
I have not been able to test on http://www.demov6.camptocamp.org/ because of https://github.com/c2corg/v6_ui/issues/894

I have then tested to upload 4 images on http://c2corgv6.demo-camptocamp.com/routes/814691/fr/calanques-les-goudes-rocher-saint-michel-d-eau-douce-le-pin-rassis => the square thumbnails are actually distorted, not cropped.
For instance:
http://c2corgv6-images.demo-camptocamp.com/active/1479144935_180623581SI.jpg
http://c2corgv6-images.demo-camptocamp.com/active/1479144935_180623581BI.jpg

**Comment by ginold** [comment]
are we speaking about "cropping" (= cut the borders of an image) or about "resizing" (= shrink proportionally)  ?

**Comment by asaunier** [comment]
Both :)
The SI images should be resized proportionally and then cropped to make them square.

**Comment by arnaud-morvan** [comment]
This is done, the dev server is not up to date
